adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m

```diff
@@ -0,0 +1,199 @@
+    if (!placementId.length) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInvalidServerParameters,
+                                                                      [NSString stringWithFormat:@"%@ cannot be nil.",GADMAdapterPanglePlacementID]
+                                                                      );
+        _loadCompletionHandler(nil, error);
+        return;
+    }
+    
+    GADPangleNetworkExtras *extras = adConfiguration.extras;
+    
+    BUAdSlot *slot = [[BUAdSlot alloc] init];
+    slot.ID = placementId;
+    slot.AdType = BUAdSlotAdTypeFeed;
+    slot.userData = extras.userDataString;
+    _nativeAd = [[BUNativeAd alloc]initWithSlot:slot];
+    _nativeAd.delegate = self;
+    [_nativeAd setAdMarkup:adConfiguration.bidResponse];
+}
+
+- (BUNativeAdRelatedView *)getRelatedView {
+    if (!_relatedView) {
+        _relatedView = [[BUNativeAdRelatedView alloc] init];
+    }
+    return _relatedView;
+}
+
+#pragma mark GADMediationNativeAd
+
+- (UIView *)mediaView {
+    return [self getRelatedView].videoAdView;
+}
+
+- (UIView *)adChoicesView {
+    return [self getRelatedView].logoADImageView;
+}
+
+- (NSString *)headline {
+    if (_nativeAd && _nativeAd.data) {
+        return _nativeAd.data.AdTitle;
+    }
+    return nil;
+}
+
+- (NSString *)body {
+    if (_nativeAd && _nativeAd.data) {
+        return _nativeAd.data.AdDescription;
+    }
+    return nil;
+}
+
+- (NSString *)callToAction {
+    if (_nativeAd && _nativeAd.data) {
+        return _nativeAd.data.buttonText;
+    }
+    return nil;
+}
+
+- (NSDecimalNumber *)starRating {
+    return [NSDecimalNumber decimalNumberWithString:[NSString stringWithFormat:@"%ld",(long)_nativeAd.data.score]];
+}
+
+- (NSString *)store {
+    return nil;
+}
+
+- (NSString *)price {
+    return nil;
+}
+
+- (NSString *)advertiser {
+    if (_nativeAd && _nativeAd.data) {
+        return _nativeAd.data.AdTitle;
+    }
+    return nil;
+}
+
+- (NSDictionary<NSString *,id> *)extraAssets {
+    if (_nativeAd) {
+        return @{BUDNativeAdTranslateKey:_nativeAd};
+    }
+    return nil;
+}
+
+#pragma mark BUNativeAdDelegate
+
+- (void)nativeAdDidLoad:(BUNativeAd *)nativeAd view:(UIView *_Nullable)view {
+    BUMaterialMeta *data = nativeAd.data;
+    // main image of the ad
+    if (data.imageAry && data.imageAry.count && data.imageAry[0].imageURL != nil){
+        _images = @[[self imageWithUrlString:data.imageAry[0].imageURL]];
+    }
+    
+    // icon image of the ad
+    if (data.icon && data.icon.imageURL != nil){
+        _icon = [self imageWithUrlString:_nativeAd.data.icon.imageURL];
+    }
+    
+    [_relatedView refreshData:nativeAd];
+    
+    if (_loadCompletionHandler) {
+        _delegate = _loadCompletionHandler(self,nil);
+    }
+}
+
+- (void)nativeAd:(BUNativeAd *)nativeAd didFailWithError:(NSError *_Nullable)error {
+    if (_loadCompletionHandler) {
+        _loadCompletionHandler(nil, error);
```

Comment:
Add logging here.

Reply:
sorry, your mean is add `NSLog(@"Pangle nativeAd:%@didFailWithError:%@ ",self,error)`?